azext_iot/iothub/job.py

```diff
@@ -0,0 +1,178 @@
+    twin_patch=None,
+    method_name=None,
+    method_payload=None,
+    method_connect_timeout=30,
+    method_response_timeout=30,
+    ttl=3600,
+    wait=False,
+    poll_interval=10,
+    hub_name=None,
+    resource_group_name=None,
+    login=None,
+):
+    from msrest.exceptions import SerializationError
+    from azext_iot.common.shared import JobType, JobStatusType
+    from azext_iot.sdk.service.models.cloud_to_device_method import CloudToDeviceMethod
+    from azext_iot.sdk.service.models.job_request import JobRequest
+
+    if (
+        job_type in [JobType.scheduleUpdateTwin.value, JobType.scheduleDeviceMethod.value]
+        and not query_condition
+    ):
+        raise CLIError(
+            "The query condition is required when job type is {} or {}. "
+            "Use query condition '*' if you need to run job on all devices.".format(
+                JobType.scheduleUpdateTwin.value, JobType.scheduleDeviceMethod.value
+            )
+        )
+
+    job_request = JobRequest(
+        job_id=job_id,
+        type=job_type,
+        start_time=start_time,
+        max_execution_time_in_seconds=ttl,
+        query_condition=query_condition,
+    )
+
+    if job_type == JobType.scheduleUpdateTwin.value:
+        if not twin_patch:
+            raise CLIError(
+                "The {} job type requires --twin-patch.".format(
+                    JobType.scheduleUpdateTwin.value
+                )
+            )
+
+        twin_patch = process_json_arg(twin_patch, argument_name="twin-patch")
+        if not isinstance(twin_patch, dict):
+            raise CLIError(
+                "Twin patches must be objects. Received type: {}".format(
+                    type(twin_patch)
+                )
+            )
+
+        # scheduleUpdateTwin job type is a force update, which only accepts '*' as the Etag.
+        twin_patch["etag"] = "*"
+        job_request.update_twin = twin_patch
+    elif job_type == JobType.scheduleDeviceMethod.value:
+        if not method_name:
+            raise CLIError(
+                "The {} job type requires --method-name.".format(
+                    JobType.scheduleDeviceMethod.value
+                )
+            )
+
+        method_payload = process_json_arg(
+            method_payload, argument_name="method-payload"
+        )
+        job_request.cloud_to_device_method = CloudToDeviceMethod(
+            method_name=method_name,
+            connect_timeout_in_seconds=method_connect_timeout,
+            response_timeout_in_seconds=method_response_timeout,
+            payload=method_payload,
+        )
+
+    target = get_iot_hub_connection_string(
+        cmd, hub_name, resource_group_name, login=login
+    )
+    service_sdk, errors = _bind_sdk(target, SdkType.service_sdk)
+
+    try:
+        job_result = service_sdk.create_job1(id=job_id, job_request=job_request)
+        if wait:
+            logger.info("Waiting for job completion state...")
+            while True:
```

Comment:
It seems there should be more control flow here. Thinks that come to mind are if `job_result` immediately returned in line 105 with an acceptable status, we shouldn't need to sleep and re-query. Also, it seems like we could end up in an indefinite sleep/poll.

Reply:
We talked about this - if the job request is accepted by the API it will always returned a queued status object. If no `--wait` is specified the status object is returned to the caller. To mitigate the infinite loop issue, we will introduce a `--poll-max-duration <minutes>` argument.

Reply:
Opted for `--poll-duration <seconds>` to keep the same units between interval and duration.